add compile_method flag and add other framework artifact types

rbavery opened the pull request:
## Description

This adds a compile_method to the MLM Asset which identifies if the model artifact has related jit or aot compilation workflows applied to it (either when saving the artifact for aot or during runtime for jit).

It also reorganizes a lot of framework specific recommendations for describing model export formats to the best-practices.md . More descriptions for Tensorflow and Keras specific export formats are added as well.

## Related Issue

<!-- If your PR refers to a related issue, link it here. -->

## Type of Change

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [ ] :books: Examples, docs, tutorials or dependencies update;
- [ ] :wrench: Bug fix (non-breaking change which fixes an issue);
- [ ] :clinking_glasses: Improvement (non-breaking change which improves an existing feature);
- [x] :rocket: New feature (non-breaking change which adds functionality);
- [ ] :boom: Breaking change (fix or feature that would cause existing functionality to change);
- [ ] :closed_lock_with_key: Security fix.

## Checklist

<!-- Mark with an `x` all the checkboxes that apply (like `[x]`) -->

- [x] I've read the [`CONTRIBUTING.md`](../CONTRIBUTING.md) guide;
- [x] I've updated the code style using `make check`;
- [ ] I've written tests for all new methods and classes that I created;
- [ ] I've written the docstring in `Google` format for all the methods and classes that I used.

fmigneault commented on README.md:
```diff
@@ -657,17 +657,24 @@ the users understand the source explicitly, although this is not strictly requir
+| `torch.save`       | A [serialized python pickle object][pytorch-save] (i.e.: `.pt`) which can represent a model or state_dict. |
+| `torch.jit.save` | A [`TorchScript`][pytorch-jit-script] model artifact obtained with one or more of the graph export options Torchscript Tracing and Torchscript Scripting.                    |
+| `torch.export.save`     | A model artifact storing an ExportedProgram obtained by [`torch.export.export`][pytorch-export] (i.e.: `.pt2`).        |
+|  `TFSavedModel`   | A [SavedModel][tf-save] from Tensorflow or Keras. |
```
Shouldn't we use something like `tf.saved_model.save` pointing at either https://www.tensorflow.org/guide/saved_model#saving_a_custom_model or https://www.tensorflow.org/api_docs/python/tf/saved_model/save to keep the naming convention consistent with a reference function?

rbavery replied:
yeah I was going back and forth on this. Keras save functions have different argument options to export in different formats. I'm not sure yet if using the function name was the best mapping to artifact type for keras and TF. the actual function exporting a specific artifact type might not be in the public api, need to check docs more thoroughly.

rbavery replied:
switched to use function names

rbavery commented on README.md:
```diff
@@ -657,17 +657,24 @@ the users understand the source explicitly, although this is not strictly requir
-| `torch.jit.script` | A model artifact obtained by [`TorchScript`][pytorch-jit-script].                    |
-| `torch.export`     | A model artifact obtained by [`torch.export`][pytorch-export] (i.e.: `.pt2`).        |
-| `torch.compile`    | A model artifact obtained by [`torch.compile`][pytorch-compile].                     |
+| `torch.save`       | A [serialized python pickle object][pytorch-save] (i.e.: `.pt`) which can represent a model or state_dict. |
```
will do

rbavery commented on README.md:
```diff
@@ -657,17 +657,24 @@ the users understand the source explicitly, although this is not strictly requir
+| `torch.export.save`     | A model artifact storing an ExportedProgram obtained by [`torch.export.export`][pytorch-export] (i.e.: `.pt2`).        |
+|  `TFSavedModel`   | A [SavedModel][tf-save] from Tensorflow or Keras. |
+| `Keras_v3`    |  Keras v3 is the [recommended format][keras-recommended] by the Tensorflow team. See this example to [save and load models][keras-example] and the update to date docs [disambiguating different save methods][keras-methods] in TF and Keras..                    |
+| `h5`    | [Keras and tf.keras model][h5] weights format, which uses HDF5.                     |
```
good point, I'll call it tf_keras_h5. it's used by both so I think including both in the name would be most descriptive.

rbavery replied:
renamed to `tf.keras.Model.save_weights`

rbavery commented on README.md:
```diff
@@ -657,17 +657,24 @@ the users understand the source explicitly, although this is not strictly requir
-| `torch.export`     | A model artifact obtained by [`torch.export`][pytorch-export] (i.e.: `.pt2`).        |
-| `torch.compile`    | A model artifact obtained by [`torch.compile`][pytorch-compile].                     |
+| `torch.save`       | A [serialized python pickle object][pytorch-save] (i.e.: `.pt`) which can represent a model or state_dict. |
+| `torch.jit.save` | A [`TorchScript`][pytorch-jit-script] model artifact obtained with one or more of the graph export options Torchscript Tracing and Torchscript Scripting.                    |
```
they are. different functions are used to trace and script all or part of a model and they can be intermixed. the load function is also the same for both. added links, thanks for the review

rbavery commented on README.md:
```diff
@@ -657,17 +657,24 @@ the users understand the source explicitly, although this is not strictly requir
+| `torch.jit.save` | A [`TorchScript`][pytorch-jit-script] model artifact obtained with one or more of the graph export options Torchscript Tracing and Torchscript Scripting.                    |
+| `torch.export.save`     | A model artifact storing an ExportedProgram obtained by [`torch.export.export`][pytorch-export] (i.e.: `.pt2`).        |
+|  `TFSavedModel`   | A [SavedModel][tf-save] from Tensorflow or Keras. |
+| `Keras_v3`    |  Keras v3 is the [recommended format][keras-recommended] by the Tensorflow team. See this example to [save and load models][keras-example] and the update to date docs [disambiguating different save methods][keras-methods] in TF and Keras..                    |
```
I renamed all these to point to specific functions and included an argument in one of them, `tf.keras.Model.export(format='tf_saved_model')` that is tensorflow specific. The other two functions always output the same artifact. Would prefer to use the high level function `tf.keras.Model.export(format='tf_saved_model')` as the artifact type since it is what users will typically interact with rather than a lower level function like `keras.saving.serialize_keras_object`

rbavery commented on README.md:
```diff
@@ -657,17 +657,24 @@ the users understand the source explicitly, although this is not strictly requir
 [pytorch-save]: https://pytorch.org/tutorials/beginner/saving_loading_models.html
+[tf-save]: https://www.tensorflow.org/guide/saved_model
+[keras-example]: https://keras.io/guides/serialization_and_saving/
+[keras-recommended]: https://www.tensorflow.org/guide/saved_model#creating_a_savedmodel_from_keras
```
done

rbavery commented:
@fmigneault I added mlm:compile_method to the JSON Schema but this seems to have broken validation of `mlm:unknown` in one of the tests. 

```
make test

――――――――――――――――――――――――――――――――――――――――――――― test_mlm_no_undefined_prefixed_field_item_properties[item_raster_bands.json] ―――――――――――――――――――――――――――――――――――――――――――――
tests/test_schema.py:61: in test_mlm_no_undefined_prefixed_field_item_properties
    with pytest.raises(pystac.errors.STACValidationError) as exc:
E   Failed: DID NOT RAISE <class 'pystac.errors.STACValidationError'>
```

I went through and added `mlm:compile_method` wherever `mlm:artifact_type` exists since I think we want validation behavior to be mostly the same for these fields. The only section I didn't add was

```
    "HasArtifactType": {
      "$comment": "Used to check the artifact type property that is required by a Model Asset annotated by 'mlm:model' role.",
      "type": "object",
      "required": [
        "mlm:artifact_type"
      ],
      "properties": {
        "mlm:artifact_type": {
          "$ref": "#/$defs/mlm:artifact_type"
        }
      }
    },
```

Any tips on what I need to change in the schema? I want it to be allowed under the asset role, if it is present I want it validated to be a string, but unlike `mlm:artifact_type` I don't want it to be required if the `mlm-model` role is there.

fmigneault commented on best-practices.md:
```diff
@@ -282,3 +282,40 @@ training process to find the "best model". This field could also be used to indi
+[pytorch-frameworks]: https://pytorch.org/docs/main/export.html#existing-frameworks
+[pytorch-jit-script]: https://pytorch.org/docs/stable/jit.html
+[pytorch-save]: https://pytorch.org/tutorials/beginner/saving_loading_models.html
+[keras-save-weights]: https://keras.io/api/models/model_saving_apis/weights_saving_and_loading/#save_weights-method
+[keras-example]: https://keras.io/guides/serialization_and_saving/
+[tf-keras-recommended]: https://www.tensorflow.org/guide/saved_model#creating_a_savedmodel_from_keras
+[keras-methods]: https://keras.io/2.16/api/models/model_saving_apis/
+[keras-model]: https://keras.io/api/models/model_saving_apis/model_saving_and_loading/
```
Not sure if they are all the correct links?
I couldn't find `tf_saved_model` under them.
I can find it here: https://keras.io/api/models/model_saving_apis/export/
But that link is not listed here.

rbavery replied:
the keras-example was incorrectly placed there thanks for catching, fixed and checked each link.

fmigneault commented on best-practices.md:
```diff
@@ -282,3 +282,40 @@ training process to find the "best model". This field could also be used to indi
+| `torch.export.save`                              | A model artifact storing an [ExportedProgram][exported-program] obtained by [`torch.export.export`][pytorch-export] (i.e.: `.pt2`).                                                                                                                                                                                                         |
+| `tf.keras.Model.save`                            | Saves a [.keras model file][keras-model], a unified zip archive format containing the architecture, weights, optimizer, losses, and metrics.                                                                                                                                                                                                |
+| `tf.keras.Model.save_weights`                    | A [.weights.h5][keras-save-weights] file containing only model weights for use by Tensorflow or Keras.                                                                                                                                                                                                                                      |
+| `tf.keras.Model.export(format='tf_saved_model')` | TF Saved Model is the [recommended format][tf-keras-recommended] by the Tensorflow team for whole model saving/loading for inference. See this example to [save and load models][keras-example] and the docs for [different save methods][keras-methods] in TF and Keras. Also available from `keras.Model.export(format='tf_saved_model')` |
```
I understand the idea of being explicit, but I wonder if that name makes it useful. 

For example, if the intention is to help create search/filter queries to find models that match this artifact type, the query definition becomes much harder to define because there will most probably be a need to escape the quotes, parenthesis and equal symbols all employed by most query languages. Wouldn't `tf.keras.Model.export` alone` be enough to hint about this use, since the `tf.` prefix is used anyway? 

I preferably would avoid hinting users toward writing definitions that look too much like directly executable code. It is preferable if there is just enough information to know which modules/functions to import, but parameters and calling them must still be done explicitly by "connecting the dots".


See also other comment about the links. Maybe not the right ones for what is described?

rbavery replied:
I agree since the tf. prefix is included, will change

rbavery commented on CHANGELOG.md:
woops, corrected. make lint-all found no issues with changelog formatting

fmigneault replied:
Oh. Maybe I need to be clearer. I mean mostly the `code` stuff that will likely not be picked up by any linter.

rbavery replied:
woops! added code ticks

rbavery commented on README.md:
```diff
@@ -667,7 +667,8 @@ In order to provide more context, the following roles are also recommended were
 | type              | string                          | The media type of the artifact (see [Model Artifact Media-Type](#model-artifact-media-type).     |
 | roles             | \[string]                       | **REQUIRED** Specify `mlm:model`. Can include `["mlm:weights", "mlm:checkpoint"]` as applicable. |
-| mlm:artifact_type | [Artifact Type](#artifact-type) | Specifies the kind of model artifact. Typically related to a particular ML framework.            |
+| mlm:artifact_type | [Artifact Type](./best-practices.md#framework-specific-artifact-types) | Specifies the kind of model artifact. Typically related to a particular ML framework. This is **REQUIRED** if the `mlm:model` role is specified.           |
```
agree, done

rbavery commented on README.md:
```diff
@@ -667,7 +667,8 @@ In order to provide more context, the following roles are also recommended were
 | roles             | \[string]                       | **REQUIRED** Specify `mlm:model`. Can include `["mlm:weights", "mlm:checkpoint"]` as applicable. |
-| mlm:artifact_type | [Artifact Type](#artifact-type) | Specifies the kind of model artifact. Typically related to a particular ML framework.            |
+| mlm:artifact_type | [Artifact Type](./best-practices.md#framework-specific-artifact-types) | Specifies the kind of model artifact. Typically related to a particular ML framework. This is **REQUIRED** if the `mlm:model` role is specified.           |
+| mlm:compile_method | string | Describes the method used to compile the ML model at either save time or runtime prior to inference. These options are mutually exclusive `["aot", "jit", null]`. |
```
agree done

rbavery commented on json-schema/schema.json:
```diff
@@ -354,7 +358,8 @@
             "anyOf": [
               {
                 "required": [
-                  "mlm:artifact_type"
+                  "mlm:artifact_type",
+                  "mlm:compile_method"
                 ]
```
I see thanks!

rbavery commented on json-schema/schema.json:
```diff
@@ -460,7 +465,18 @@
+        "tf.saved_model.export(format='tf_saved_model')"
+      ]
+    },
+    "mlm:compile_method": {
+      "type": "string",
+      "minLength": 1,
```
yes that's the intention. I think it'd be burdensome to require setting null explicitly here.

fmigneault replied:
Ok great. I suggest the set of values then omits the `null`, and that is mentioned only in the type as `Compile Method | null`. This will match the representation used by other fields.

fmigneault commented on README.md:
```diff
@@ -667,7 +667,8 @@ In order to provide more context, the following roles are also recommended were
 | roles             | \[string]                       | **REQUIRED** Specify `mlm:model`. Can include `["mlm:weights", "mlm:checkpoint"]` as applicable. |
-| mlm:artifact_type | [Artifact Type](#artifact-type) | Specifies the kind of model artifact. Typically related to a particular ML framework.            |
+| mlm:artifact_type | [Artifact Type](./best-practices.md#framework-specific-artifact-types) | Specifies the kind of model artifact, any string is allowed. Typically related to a particular ML framework, see [Best Practices - Framework Specific Artifact Types](./best-practices.md#framework-specific-artifact-types) for **RECOMMENDED** values. This field is **REQUIRED** if the `mlm:model` role is s ...
+| mlm:compile_method | [Compile Method](#compile-method) | null | Describes the method used to compile the ML model either when the model is saved or at model runtime prior to inference. |
```
Missing the `\|` to escape within the table.

rbavery replied:
woops, good catch!
